runtime/health.go

```diff
@@ -58,5 +50,183 @@ func (r *Runtime) Health(ctx context.Context) (*Health, error) {
 }
 
 func (r *Runtime) InstanceHealth(ctx context.Context, instanceID string) (*InstanceHealth, error) {
-	return r.registryCache.instanceHealth(ctx, instanceID)
+	res := &InstanceHealth{}
+	// check repo error
+	err := r.pingRepo(ctx, instanceID)
+	if err != nil {
+		res.Repo = err.Error()
+	}
+
+	ctrl, err := r.Controller(ctx, instanceID)
+	if err != nil {
+		res.Controller = err.Error()
+		return res, nil
+	}
+
+	parser, err := ctrl.Get(ctx, GlobalProjectParserName, false)
+	if err != nil {
+		return nil, err
+	}
+	res.ParseErrCount = len(parser.GetProjectParser().State.ParseErrors)
+
+	cachedHealth, _ := r.cachedInstanceHealth(ctx, ctrl.InstanceID, ctrl.catalog.version)
+	// set to true if any of the olap engines can be scaled to zero
+	var canScaleToZero bool
+
+	// check OLAP error
+	olap, release, err := r.OLAP(ctx, instanceID, "")
+	if err != nil {
+		res.OLAP = err.Error()
+	} else {
+		mayBeScaledToZero := olap.MayBeScaledToZero(ctx)
+		canScaleToZero = canScaleToZero || mayBeScaledToZero
+		if cachedHealth != nil && mayBeScaledToZero {
+			res.OLAP = cachedHealth.OLAP
+		} else {
+			err = r.pingOLAP(ctx, olap)
+			if err != nil {
+				res.OLAP = err.Error()
+			}
+		}
+		release()
+	}
+
+	// run queries against metrics views
+	resources, err := ctrl.List(ctx, ResourceKindMetricsView, "", false)
+	if err != nil {
+		return nil, err
+	}
+	res.MetricsViews = make(map[string]metricsViewHealth, len(resources))
+	for _, mv := range resources {
+		if mv.GetMetricsView().State.ValidSpec == nil {
+			if mv.Meta.ReconcileError != "" {
+				res.ReconcileErrCount++
+			}
+			continue
+		}
+		olap, release, err := r.OLAP(ctx, instanceID, mv.GetMetricsView().State.ValidSpec.Connector)
+		if err != nil {
+			res.MetricsViews[mv.Meta.Name.Name] = metricsViewHealth{err: err.Error()}
+			release()
+			continue
+		}
+		mayBeScaledToZero := olap.MayBeScaledToZero(ctx)
+		canScaleToZero = canScaleToZero || mayBeScaledToZero
+		release()
+
+		// only use cached health if the OLAP can be scaled to zero
+		if cachedHealth != nil && mayBeScaledToZero {
+			mvHealth, ok := cachedHealth.MetricsViews[mv.Meta.Name.Name]
+			if ok && mvHealth.Version == mv.Meta.StateVersion {
+				res.MetricsViews[mv.Meta.Name.Name] = mvHealth
+				continue
+			}
+		}
+		_, err = r.Resolve(ctx, &ResolveOptions{
+			InstanceID:         ctrl.InstanceID,
+			Resolver:           "metricsview_time_range",
+			ResolverProperties: map[string]any{"metrics_view": mv.Meta.Name.Name},
+			Args:               map[string]any{"priority": 10},
+			Claims:             &SecurityClaims{SkipChecks: true},
+		})
+
+		mvHealth := metricsViewHealth{Version: mv.Meta.StateVersion}
+		if err != nil {
+			mvHealth.err = err.Error()
+		}
+		res.MetricsViews[mv.Meta.Name.Name] = mvHealth
+	}
+
+	if !canScaleToZero {
+		return res, nil
+	}
```

Comment:
The reason I was not storing this for all OLAP engines was because it doesn't handle external tables very well. Till the metrics view is not updated it will keep reporting healthy even if the underlying OLAP engine goes down.

We can improve this logic in subsequent PRs and then we can cache it for all.
1. For other OLAP engines we can invalidate the cache after a TTL.
2. For clickhouse we can invalidate the cache if we are able to retrieve the actual status of the service.
Please let me me know what do you think.

Reply:
Though I think we can store but not use it here which means it will be refreshed after every health check.